ghcide/src/Development/IDE/Main.hs

```diff
@@ -289,9 +289,14 @@ testing recorder logger =
       , argsIdeOptions = ideOptions
       }
 
-
-defaultMain :: Recorder (WithPriority Log) -> Arguments -> IO ()
-defaultMain recorder Arguments{..} = withHeapStats (cmapWithPrio LogHeapStats recorder) fun
+defaultMain
+    :: Recorder (WithPriority Log)
+    -> Maybe (MVar (LSP.LanguageContextEnv Config))
+    -- ^ Variable to be filled with the LSP environment, useful for tools that need this outside
+    -- the scope of runLanguageServer
+    -> Arguments
+    -> IO ()
+defaultMain recorder lspEnvVar Arguments{..} = withHeapStats (cmapWithPrio LogHeapStats recorder) fun
```

Comment:
The `MVar` is an implementation detail, better hide it behind a callback, or a hook to ran an `Lsp` action on initialisation.

And please move it to the `Arguments` record, to keep the API simple and minimise the breakage

Reply:
I wasn't sure about whether to put it in `Arguments`, in some places there seems to be an assumption that it reflects CLI arguments. But yeah, it seems better in there.

Reply:
I ended up putting it in the `Arguments` record for `ghcide` but not HLS: the HLS `Arguments` are very tightly coupled to argument parsing, expect to be `Show`-able etc, so it really didn't fit.